File: score.py
from globals import *

__all__ = ['killed_by', 'win', 'quit']

SCOREFILE = "scores"

def killed_by(monster, other):
    #signal(SIGINT, SIG_IGN)
    
    if other != QUIT:
        rogue.gold = rogue.gold * 9 / 10
    
    if other == HYPOTHERMIA:
        buf = "died of hypothermia"
    elif other == STARVATION:
        buf = "died of starvation"
    elif other == QUIT:
        buf = "quit"
    else:
        buf = "killed by "
        name = monster_names[ord(monster.ichar) - ord('A')]
        if  is_vowel(name):
            buf += "an "
        else:
            buf += "a "
        buf += name
    
    buf += " with %d gold" % rogue.gold
    message(buf, 0)
    message("", 0)
    score(monster, other)
    
def win():
    rogue.armor = None
    rogue.weapon = None
    
    clear()
    mvaddstr(10, 11, "@   @  @@@   @   @      @  @  @   @@@   @   @   @")
    mvaddstr(11, 11, " @ @  @   @  @   @      @  @  @  @   @  @@  @   @")
    mvaddstr(12, 11, "  @   @   @  @   @      @  @  @  @   @  @ @ @   @")
    mvaddstr(13, 11, "  @   @   @  @   @      @  @  @  @   @  @  @@")
    mvaddstr(14, 11, "  @    @@@    @@@        @@ @@    @@@   @   @   @")
    mvaddstr(17, 11, "Congratulations,  you have  been admitted  to  the")
    mvaddstr(18, 11, "Fighter's Guild.   You return home,  sell all your")
    mvaddstr(19, 11, "treasures at great profit and retire into comfort.")
    message("", 0);
    message("", 0);
    id_all()
    sell_pack()
    score(None, WIN)
    
def quit():
    message("really quit?", 1)
    if getchar() != 'y':
        check_message()
        return
    check_message()
    killed_by(None, QUIT)
    
def score(monster, other):
    # todo loop in case the scores file cannot be accessed/created
    put_scores(monster, other)
    
def put_scores(monster, other):
    scores = [""] * 10
    
    f = open(SCOREFILE, "a+") # read an existing file or create a new one
    f.seek(0)
    
    rank = 10
    dont_insert = 0
    i = 0
    while i < 10:
        #L:
        scores[i] = f.readline()
        if scores[i] == "":
            break
        if len(scores[i]) < 18:
            message("error in score file format", 1)
            cleanup("sorry, score file is out of order")
        if ncmp(scores[i][16:], g.player_name):
            s = int(scores[i][8:16])
            if s <= rogue.gold:
                #goto L
                continue
            dont_insert = 1
        i += 1
    
    #if dont_insert: goto DI
    if not dont_insert:
        for j in range(i):
            if rank > 9:
                s = int(scores[j][8:16])
                if s <= rogue.gold:
                    rank = j
        
        if i == 0:
            rank = 0
        elif i < 10 and rank > 9:
            rank = i
        if rank <= 9:
            insert_score(scores, rank, i, monster, other)
            if i < 10:
                i += 1
                
        f.truncate(0)

    #DI:
    clear()
    mvaddstr(3, 30, "Top  Ten  Rogueists")
    mvaddstr(8, 0, "Rank    Score   Name")
    
    #signal(SIGQUIT, SIG_IGN)
    #signal(SIGINT, SIG_IGN)
    #signal(SIGHUP, SIG_IGN)
    
    for j in range(i):
        if j == rank:
            standout()
        scores[j] = "%2d" % (j + 1) + scores[j][2:]
        mvaddstr(j + 10, 0, scores[j])
        if rank < 10:
            f.write(scores[j])
        if j == rank:
            standend()

    refresh()
    f.close()
    
    wait_for_ack("")
    
    clean_up("")

def insert_score(scores, rank, n, monster, other):
    for i in range(n - 1, rank - 1, -1):
        if i < 9:
            scores[i + 1] = scores[i]
    buf = "%2d      %5d   %s: " % (rank + 1, rogue.gold, g.player_name)
    
    if other == HYPOTHERMIA:
        buf += "died of hypothermia"
    elif other == STARVATION:
        buf += "died of starvation"
    elif other == QUIT:
        buf += "quit"
    elif other == WIN:
        buf += "a total winner"
    else:
        buf += "killed by "
        name = monster_names[ord(monster.ichar) - ord('A')]
        if is_vowel(name):
            buf += "an "
        else:
            buf += "a "
        buf += name
    buf += " on level %d " % g.max_level
    if other != WIN and g.has_amulet:
        buf += "with amulet"
    buf += "\n"
    scores[rank] = buf

def is_vowel(ch):
    return ch in "aeiou"

def sell_pack():
    rows = 2

    clear()
    
    obj = rogue.pack.next_object
    while obj:
        mvaddstr(1, 0, "Value      Item")
        if obj.what_is != FOOD:
            obj.identified = 1
            val = get_value(obj)
            rogue.gold += val
            
            if rows < SROWS:
                mvadstr(row, 0, "%5d      %s" % (val, get_description(obj)))
                row += 1
        obj = obj.next_object
    refresh()
    message("", 0)
    
def get_value(obj):
    k = obj.which_kind
    if k == WEAPON:
        val = id_weapons[k].value
        if k == ARROW or k == SHURIKEN:
            val *= obj.quantity
        val += obj.damage_enchantment * 85
        val += obj.to_hit_enchantment * 85
    elif k == ARMOR:
        val = id_armors[k].value
        val += obj.damage_enchantment * 75
        if obj.is_protected:
            val += 200
    elif k == WAND:
        val = id_wands[k].value * obj.clasz
    elif k == SCROLL:
        val = id_scrolls[k].value * obj.quantity
    elif k == POTION:
        val = id_potions[k].value * obj.quantity
    elif k == AMULET:
        val = 5000
    else:
        val = 0
    return max(val, 10)

def id_all():
    for i in range(SCROLLS):
        id_scrolls[i].id_status = IDENTIFIED
    for i in range(WEAPONS):
        id_weapons[i].id_status = IDENTIFIED
    for i in range(ARMORS):
        id_armors[i].id_status = IDENTIFIED
    for i in range(WANDS):
        id_wands[i].id_status = IDENTIFIED
    for i in range(POTIONS):
        id_potions[i].id_status = IDENTIFIED

def ncmp(s1, s2):
    return s1[:s1.index(":")] == s2

File: flrogue/lib/rogue/score.dart
import 'dart:io' if (dart.library.js) 'web_io.dart';

import 'globals.dart';
import 'init.dart';
import 'message.dart';
import 'object.dart';
import 'ui.dart';

const String scoreFile = "scores";

Future<void> killedBy(GameObject? monster, DeathCause other) async {
  if (other != DeathCause.quit) {
    rogue.gold = rogue.gold * 9 ~/ 10;
  }

  String buf;
  if (other == DeathCause.hypothermia) {
    buf = "died of hypothermia";
  } else if (other == DeathCause.starvation) {
    buf = "died of starvation";
  } else if (other == DeathCause.quit) {
    buf = "quit";
  } else if (other != DeathCause.win) {
    buf = "killed by ";
    String name = monsterNames[monster!.ichar.ascii - 'A'.ascii];
    if (_isVowel(name[0])) {
      buf += "an ";
    } else {
      buf += "a ";
    }
    buf += name;
  } else {
    buf = "";
  }

  buf += " with ${rogue.gold} gold";
  await message(buf);
  await message("");
  await score(monster, other);

  ui.end();
  exit(0);
}

Future<void> win() async {
  rogue.armor = null;
  rogue.weapon = null;

  ui.clearScreen();
  ui.move(10, 11).write("@   @  @@@   @   @      @  @  @   @@@   @   @   @");
  ui.move(11, 11).write(" @ @  @   @  @   @      @  @  @  @   @  @@  @   @");
  ui.move(12, 11).write("  @   @   @  @   @      @  @  @  @   @  @ @ @   @");
  ui.move(13, 11).write("  @   @   @  @   @      @  @  @  @   @  @  @@");
  ui.move(14, 11).write("  @    @@@    @@@        @@ @@    @@@   @   @   @");
  ui.move(17, 11).write("Congratulations,  you have  been admitted  to  the");
  ui.move(18, 11).write("Fighter's Guild.   You return home,  sell all your");
  ui.move(19, 11).write("treasures at great profit and retire into comfort.");

  await message("");
  await message("");
  _idAll();
  await _sellPack();
  await score(null, DeathCause.win);

  ui.end();
  exit(0);
}

Future<void> quit() async {
  await message("really quit?", true);
  String ch = await ui.getchar();
  if (ch != 'y') {
    checkMessage();
    return;
  }

  checkMessage();
  await killedBy(null, DeathCause.quit);
}

Future<void> score(GameObject? monster, DeathCause other) async {
  await putScores(monster, other);
}

Future<void> putScores(GameObject? monster, DeathCause other) async {
  List<String> scores = List.filled(10, "");

  File f;
  try {
    f = File(scoreFile);
    if (!f.existsSync()) {
      f.createSync();
    }
  } catch (e) {
    await message("Cannot access score file: $e", true);
    return;
  }

  int rank = 10;
  bool dontInsert = false;
  int i = 0;

  try {
    List<String> lines = f.readAsLinesSync();
    while (i < 10 && i < lines.length) {
      scores[i] = lines[i];

      if (scores[i].length < 18) {
        await message("error in score file format", true);
        cleanUp("sorry, score file is out of order");
      }

      if (_ncmp(scores[i].substring(16), playerName)) {
        int s = int.parse(scores[i].substring(8, 16));
        if (s > rogue.gold) {
          dontInsert = true;
        }
      }

      i++;
    }

    if (!dontInsert) {
      for (int j = 0; j < i; j++) {
        if (rank > 9) {
          int s = int.tryParse(scores[j].substring(8, 16)) ?? 0;
          if (s <= rogue.gold) {
            rank = j;
          }
        }
      }

      if (i == 0) {
        rank = 0;
      } else if (i < 10 && rank > 9) {
        rank = i;
      }

      if (rank <= 9) {
        _insertScore(scores, rank, i, monster, other);
        if (i < 10) {
          i += 1;
        }
      }

      // Clear file before writing new scores
      f.writeAsStringSync("");
    }

    ui.clearScreen();
    ui.move(3, 30).write("Top  Ten  Rogueists");
    ui.move(8, 0).write("Rank    Score   Name");

    for (int j = 0; j < i; j++) {
      bool isRankHighlighted = j == rank;

      scores[j] = "${j + 1}".padLeft(2) + scores[j].substring(2);
      ui.move(j + 10, 0);

      if (isRankHighlighted) {
        ui.write(scores[j], inverse: true);
      } else {
        ui.write(scores[j]);
      }

      if (rank < 10) {
        f.writeAsStringSync('${scores[j]}\n', mode: FileMode.append);
      }
    }

    ui.refresh();
  } catch (e) {
    await message("Error processing scores: $e", true);
  }

  await waitForAck("");
  cleanUp("");
}

void _insertScore(
  List<String> scores,
  int rank,
  int n,
  GameObject? monster,
  DeathCause other,
) {
  for (int i = n - 1; i >= rank; i--) {
    if (i < 9) {
      scores[i + 1] = scores[i];
    }
  }

  String buf =
      "${"${rank + 1}".padLeft(2)}    ${"${rogue.gold}".padLeft(7)}   $playerName: ";

  if (other == DeathCause.hypothermia) {
    buf += "died of hypothermia";
  } else if (other == DeathCause.starvation) {
    buf += "died of starvation";
  } else if (other == DeathCause.quit) {
    buf += "quit";
  } else if (other == DeathCause.win) {
    buf += "a total winner";
  } else {
    buf += "killed by ";
    String name = monsterNames[monster!.ichar.ascii - 'A'.ascii];
    if (_isVowel(name[0])) {
      buf += "an ";
    } else {
      buf += "a ";
    }
    buf += name;
  }

  buf += " on level $maxLevel ";
  if (other != DeathCause.win && hasAmulet) {
    buf += "with amulet";
  }

  scores[rank] = buf;
}

bool _isVowel(String ch) {
  return "aeiou".contains(ch.toLowerCase());
}

bool _ncmp(String s1, String s2) {
  return s1.split(':')[0].trim() == s2;
}

void _idAll() {
  for (int i = 0; i < ScrollType.values.length; i++) {
    idScrolls[i].idStatus = IdStatus.identified;
  }
  for (int i = 0; i < WeaponType.values.length; i++) {
    idWeapons[i].idStatus = IdStatus.identified;
  }
  for (int i = 0; i < ArmorType.values.length; i++) {
    idArmors[i].idStatus = IdStatus.identified;
  }
  for (int i = 0; i < WandType.values.length; i++) {
    idWands[i].idStatus = IdStatus.identified;
  }
  for (int i = 0; i < PotionType.values.length; i++) {
    idPotions[i].idStatus = IdStatus.identified;
  }
}

Future<void> _sellPack() async {
  int rows = 2;

  ui.clearScreen();

  for (GameObject obj in rogue.pack) {
    ui.move(1, 0).write("Value      Item");

    if (obj.whatIs != Cell.food) {
      obj.identified = 1;
      int val = _getValue(obj);
      rogue.gold += val;

      if (rows < sRows) {
        ui
            .move(rows, 0)
            .write("${"$val".padLeft(5)}      ${getDescription(obj)}");
        rows += 1;
      }
    }
  }

  ui.refresh();
  await message("");
}

int _getValue(GameObject obj) {
  int k = obj.whichKind;
  int val = 0;

  if (k == Cell.weapon) {
    val = idWeapons[k].value;
    if (k == WeaponType.arrow.index || k == WeaponType.shuriken.index) {
      val *= obj.quantity;
    }
    val += obj.damageEnchantment * 85;
    val += obj.toHitEnchantment * 85;
  } else if (k == Cell.armor) {
    val = idArmors[k].value;
    val += obj.damageEnchantment * 75;
    if (obj.isProtected != 0) {
      val += 200;
    }
  } else if (k == Cell.wand) {
    val = idWands[k].value * obj.clasz;
  } else if (k == Cell.scroll) {
    val = idScrolls[k].value * obj.quantity;
  } else if (k == Cell.potion) {
    val = idPotions[k].value * obj.quantity;
  } else if (k == Cell.amulet) {
    val = 5000;
  }

  return val > 10 ? val : 10;
}
